lib/utils.js

```diff
@@ -116,11 +117,26 @@ export function mkdir(dirName: string): Promise<void> {
   })
 }
 
-export function rmdir(dirName: string): Promise<void> {
-  return new Promise((resolve, reject) => {
-    // The keybase service will have made some files here; we want to clean them all out
-    // Node doesn't have a recursive dir clean, but.
-    console.log(`TODO: implement recurseive rmdir for tmp directory ${dirName}`)
-    resolve()
-  })
+export async function rmdirRecursive(dirName: string): Promise<void> {
+  const fsLstat = promisify(fs.lstat)
```

Comment:
Since this code is mostly synchronous aside from the `await`s, would it make sense to use the synchronous node functions, e.g., `fs.lstatSync`?

Reply:
But isn't this how `await fsLstat` works?  The code will block until `fsLstat` resolves or throws an error due to the `await` (I think). 

It seems like the synchronousness of this function is kinda unavoidable: we need to wait on `lstat`  to give us `dirStat` before we can call `readdir` to get `entry` before we can call `stat` on entry, etc.

Reply:
`await fsLstat` is syntactic sugar for `somePromise.then(result => ...)`

It looks imperative but it isn’t. Other execution can occur (in other promises or functions) until the time when `somePromise` resolves.

Reply:
^^ just came back from doing a lot of reading about async/await nitty gritty, agree with @thebearjew. We don't want to use the synchronous versions here

Reply:
👍 Awesome. It's confusing at first because `await` is only blocking within the execution of a function prefixed with `async`.

I always imagine the code that follows a line with `await` to be magically nested into the body of `somePromise.then(result => { ...  rest of my code ...})`

Reply:
this is basically what we've been doing with max's iced coffee script. It's nice to see it finally in javascript after all these years. Async code that looks imperative. I'm still figuring out the JS version. In iced you can do nice things such as:

```
for thing in things
   await doSomething thing, defer result
```
which would `doSomething` a bunch of times, and of course be non-blocking. But if you wanted to do them all concurrently you'd just move the await outside:

```
await
  for thing,i in things
    doSomething thing, defer result[i]
```

similarly:

```
await doOnething defer result1
await doSecondThing defer result2
console.log(result1, result2)
```
does them without blocking the JS thread from doing other work, but does them sequentially, and

```
await
  doOnething defer result1
  doSecondThing defer result2
console.log(result1, result2)
```
does them concurrently.

this is all very easy to refactor. I wish it were as clean in Javascript's version...it seems not to be as far as I can tell. I think you've gotta move them into a `Promise.all` call with an array of the promises and then await that, which seems ugly. But I'm still learning, so I could be wrong. In any case this, this is way better than callback hell, and I think better than looking at promises directly which aren't much better. Enjoying learning this.